[#45083] Update look of activity items in activity module for project and work packages

mrmir opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/45083

mrmir committed:
Add basic html scaffolding for project activity list

Commit:
Update spacings in project activity list

Co-authored-by: Christophe Bliard <c.bliard@openproject.com>
Co-authored-by: Mir Bhatia <m.bhatia@openproject.com>

Commit:
Remove unused BEM element classes

cbliard committed:
Add activity description

cbliard committed:
WIP: format project journal attributes for activity

cbliard committed:
wip: add test stub for project attributes testing

mrmir committed:
WIP: Update active status journal formatter

mrmir committed:
Add active status journal formatter test

cbliard committed:
Prevent N+1 issue when displaying project activities

The journals relations are eager loaded to display the changes in the
activities page.

cbliard committed:
Fix tests

cbliard committed:
refactor: Remove double negative

cbliard committed:
Cache resources fetched when rendering journal details

The lifetime of this cache is the request lifetime (using RequestStore).

mrmir committed:
Update the look of activity items

mrmir committed:
WIP: Add scaffolding for project attributes activity spec

mrmir committed:
WIP: Add project journal attribute for template

mrmir committed:
Add project journal attribute for template

cbliard committed:
perf: Save 7% time for WorkPackage activity json rendering

Looking up class names each time something is rendered adds up. Looking
up all HTML formatter filters takes 0.5ms. Done once for each comment,
it takes 50ms for 100 comments.

This makes also the code easier to read and discover.

mrmir committed:
Remove inefficient event.journal method

mrmir committed:
Remove stray binding.pry

mrmir committed:
WIP: Add project journal attribute for identifier

cbliard committed:
Reuse PlaintextFormatter for ProjectJournal identifier field

mrmir committed:
Add project journal attribute for name

mrmir committed:
Add project journal attribute for description

mrmir commented on app/controllers/journals_controller.rb:
```diff
@@ -98,7 +98,7 @@ def valid_field?(field)
 
   def valid_diff?
     valid_field?(params[:field]) &&
-      @journal.journable.instance_of?(WorkPackage)
```
@cbliard do you foresee any problems with this? It was the easiest way to get diff to play nice for project activity imo

cbliard replied:
@mrmir This condition seems strange as it  restricts the usage of diffing things to `WorkPackage`, and thus reduce reusability.
I inspected the history of the file with `git log app/controllers/journals_controller.rb` to find why this condition has been introduced.
It has been added in cc2afee99c. The commit message states this:

>    fix controller response
>   
>   As the diff action claims to be more generic, react with a 404 instead
>   of a 403 whenever the journable type is not a WorkPackage.
>   
>   This makes the the use of the diff action less generic, as a user will
>   still need the view_work_packages permission to access the diff action.

That's interesting! It looks like this endpoint requires the `view_work_packages` permission to work. So I tried disabling the work package tracking module and it successfully lead to an error: in this configuration, displaying the diff of a project description change will lead to a 403.

This should be fixed as it should work even if the work package tracking module is disabled.

cbliard replied:
I handled it in 3c4742792b.

mrmir committed:
Add project journal attribute for visibility

mrmir committed:
Add project journal attribute for parent

cbliard committed:
Fix Zeitwerk::NameError error

lib/open_project/journal_formatter/parent.rb was not properly defining OpenProject::JournalFormatter::Parent class as the code was commented out.

Rubocop linting errors should still be fixed.

cbliard committed:
Fix error on require

For some reason, Zeitwerk does not like loading the file from itself. Excluding it from the list of files to require fixes it.

cbliard committed:
Optimize query to fetch events for WorkPackageJournals

cbliard committed:
perf: Skip subprojects sql condition when no subprojects

cbliard committed:
perf: Reduce number of allocated objects

cbliard committed:
refactor: Remove dead code

cbliard committed:
Fix failing test

cbliard committed:
Fix some rubocop linting errors

cbliard committed:
Refactor with better names and simplified logic

cbliard committed:
Fix inversed templated formatter

cbliard committed:
Hide non-rendered activity details

cbliard committed:
Render custom fields in Project activity details

cbliard committed:
Remove title attribute on i tag

HTML attribute 'title' does not help accessibility, making it useless and repetitive in this case.

See https://www.a11yproject.com/posts/title-attributes/

cbliard committed:
Render parent field in Project activity details

Reuse NamedAssociation formatter instead of defining a new one.

cbliard committed:
Render templated field in Project activity details as specified

cbliard committed:
Add i18n for project activity details

cbliard committed:
refactor: Use generic formatter for budget_id

The :cost_association formatter used a Proc to get the Budget record
from the WorkPackage association and display its name. That's exactly
what JournalFormatter::NamedAssociation does, so better use it instead
and remove dead code: remove the JournalFormatter::Proc and the code
associated with it.

cbliard committed:
Hide project name for project activity items

cbliard committed:
Use project icon on activity list of user page

cbliard committed:
refactor: Shorten line length

cbliard committed:
Handle finer permissions for journals/diff endpoint

As description of both WorkPackage and Project journal can be diffed,
the needed permissions have to be distinct: view_project for Project
description diff, and view_work_packages for WorkPackage description
diff.

cbliard committed:
Fix tests

ulferts commented on app/models/project.rb:
```diff
@@ -210,10 +219,16 @@ def project_condition(with_subprojects)
   end
 
+  def has_subprojects?
+    rgt - lft > 1
```
This can be simplified to:
```
!leaf?
```

cbliard replied:
Done. Thanks

ulferts commented on db/migrate/20230130134630_add_index_for_work_package_journals_project_id.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexForWorkPackageJournalsProjectId < ActiveRecord::Migration[7.0]
+  def change
+    add_index :work_package_journals, :project_id
```
Just an assumption, but I would guess that a similar index should be added to the other data journal classes e.g. `news_journals` and `time_entry_journals`

cbliard replied:
Thanks. Added

ulferts commented on app/models/work_package/journalized.rb:
```diff
@@ -77,33 +77,24 @@ def self.event_url
-        record = association.class_name.constantize.find_by(id: value.to_i)
-        record&.subject
-      end
-    end
```
Thanks for cleaning up 👍

ulferts commented on app/controllers/journals_controller.rb:
```diff
@@ -91,14 +93,23 @@ def find_journal
+      when 'WorkPackage' then :view_work_packages
+      when 'Project' then :view_project
+      end
+    do_authorize(permission)
```
Most (did not check all) journable classes have a `visible?` method. So it would be an option to use 

```
deny_access unless @journal.journable.visible?
```

This has the benefit of avoiding the list of permissions.

ulferts replied:
On the other hand, if it is kept like this, then I'd think the [check for the `journable_type`](https://github.com/opf/openproject/pull/11925/files#diff-341df1934319c300106bd8dc743cbb8c820b1731a1d4e67783ea170fd808011fR112) could be removed all-together. Or the checks are combined.

cbliard replied:
Thanks, I kept the code like this (without `visible?`) and removed the check for the `journable_type` (and a few other things along it)

cbliard replied:
in 0e634e24d911a36df0407eba52d0fcc0be3b1146

ulferts commented on app/views/users/show.html.erb:
```diff
@@ -87,9 +87,7 @@ See COPYRIGHT and LICENSE files for more details.
-                  <% event_type = e.event_type == 'meeting' ? 'meetings' : e.event_type %>
-                  <% event_type = e.event_type == 'budget' ? 'budget' : event_type %>
-                  <%= icon_wrapper("icon-context icon-#{event_type}", e.event_name) %>
+                  <%= icon_wrapper("icon-context icon-#{event_icon(e)}", e.event_name) %>
```
👍

ulferts reviewed (changes requested):
I honestly only have nuisances to nag about. Overall, a job well done.

cbliard committed:
refactor: Simplify logic to validate params

cbliard commented:
> I honestly only have nuisances to nag about. Overall, a job well done.

Thanks! I have addressed your comments. I'll merge it once the CI is green.
